### DISCUSSION TITLE
How to use an existing Vue directive in Quasar - help needed

### DISCUSSION THREAD

**Discussion by mickey-bimlive** [quasarframework/quasar#12536] [General - Components / Directives / etc] [answered]
I would like to use a relatively simple  Vue directive **VueAuthImage** from NPM/GitHub (a quite popular package) in my Quasar app.
My app is still on Vue 2, Quasar 1.15.21.

Here is the Vue directive VueAuthImage I would like to use: https://github.com/xzag/vue-auth-image 

I installed it, and followed the instructions, which are for Vue, not Quasar:
```
import Vue from "vue";
import VueAuthImage from "vue-auth-image";
Vue.use(VueAuthImage);
```

When I try to use that directive within my template through 
```
<div>
  <img v-auth-image="https://my.domain.com">
</div>
```
it throws strange errors. It looks like the above steps did not register that Vue component correctly to be used in my Quasar app.

What am I doing wrong? I'm not familiar with Vue directives at all. Help much appreciated!

**Comment by MilosPaunovic** [comment]
Which exact error is it throwing?

Where have you registered it, in the boot file? If so, have you included that boot file into `quasar.conf.js`?

**Comment by mickey-bimlive** [comment]
The error is basic, I don't remember it right now, I had to undo the changes since it failed.
But no, I don't have a boot file for it. I don't think it comes with one. What is the code for such a boot file?

**Comment by yusufkandemir** [comment, marked as answer]
> (a quite popular package)

It has 5 stars on GitHub and the last update was 2 years ago, so I wouldn't call it quite popular. This might also indicate it might not be really stable in all cases. So, it might be good to keep that in mind.

Here is how you would create a boot file to initialize the plugin:
```js
// src/boot/vue-auth-image.js
import { boot } from "quasar/wrappers";
import VueAuthImage from "vue-auth-image";

export default boot(({ Vue }) => {
  Vue.use(VueAuthImage);
});
```

then you would need to add it to `quasar.conf.js > boot` like:
```js
boot: [
  // ...
  'vue-auth-image',
]
```

After doing these steps, if it still fails, it is probably related to the library. This means you can try your luck by opening an issue, trying to find another package, or forking the package and fixing it yourself.


The example on the README and the one inside the example directory are different. The correct one is inside the example directory:
https://github.com/xzag/vue-auth-image/blob/1ec0498927cac90b085b2729fb10994dab9e8d66/example/index.html#L15
```html
<img v-auth-image="'https://raw.githubusercontent.com/vuejs/art/master/vue-badge.png'">
```
Notice the single quotes(') inside double quotes(").

**Reply by yusufkandemir** [reply]
@mickey58github I updated my answer. Also, please use replies instead of writing new "answer"s.

**Reply by mickey-bimlive** [reply]
Thanks - see below, got it working

**Reply by mickey-bimlive** [reply]
@yusufkandemir - I declared victory too early...it works with quasar dev --modern, but it causes my production build on Heroku to break:
Module not found: Error: Can't resolve 'vue-auth-image' in '/tmp/build_fd03e8d8/src/boot'

Can you tell how to deal with such directives in a prod build? In my local env, I installed the code through npm install vue-auth-image --save.
But strangely, I don't see it in my package-json. That's probably causing the build break on Heroku...

**Reply by MilosPaunovic** [reply]
You need to separate `--save` part from the name of the package in  `npm install vue-auth-image--save`.

**Reply by mickey-bimlive** [reply]
@MilosPaunovic - the --save without the blank was just a typo in my original reply (corrected), definitely not the cause of the build break on Heroku...

**Reply by mickey-bimlive** [reply]
@MilosPaunovic  - I repeated the install, and now my package.json got updated....with that, the Heroku build works. Thanks for your hint to look at the install again....it was not the missing blank but something else that must have spoiled my environment.

**Comment by mickey-bimlive** [comment]
Thanks for the boot file, I'll try that out later today!

**Comment by mickey-bimlive** [comment]
@yusufkandemir - I tried it with the boot file per your instructions. But the problem is basically the same as before.

First, VS Code, in the boot file shows "Import Vue from 'vue'" greyed out, as unused, but I ignored that.

Next, I tried to use the VueAuthImage directive in my template as follows (per the docs):
`<img v-auth-image='google.com'>`

(Side note) If I try instead:
`<img v-auth-image='https://google.com'>`
it gives me a compile problem, unexpected token at the https:// part of the string, although this would be the normal use case (url) for it...

Next, if I try it without the "https://" prefix, at runtime it throws those errors as without the boot file:
```
[Vue warn]: Property or method "google" is not defined on the instance but referenced during render. 
[Vue warn]: Error in render: "TypeError: Cannot read properties of undefined (reading 'com')"
```

I can inspect the source code in the browser as follows:
 ```
...
 directives: [
  {
    name: "auth-image",
    rawName: "v-auth-image",
    value: _vm.google.com,
    expression: "google.com"
  }
]
```
So, something is still broken with my use of it...or this directive has general problems, as you mentioned as possible.

**Comment by mickey-bimlive** [comment]
Success: I got it working, based on a hint on https://github.com/xzag/vue-auth-image/issues/1
It requires double quotes for the URL prop/value, for whatever reason:
`<img v-auth-image="'https://dummyimage.com/100'">`
@yusufkandemir - thanks for your help!

However, it works only with quasar - dev, and causes a build break in production build...
